### ISSUE TITLE
MoveFor() oddly not completing

### ISSUE THREAD

**Issue by gcl8a** [WPIRoboticsEngineering/RBE1001Lib#14]
We're having intermittent trouble with Motor::MoveFor(). Before, it would check if the motor had reached 90% of it's motion and then check to see if it had stopped. The first check is to make sure the motor has started (and moved most of the way); the second to detect the end (even if it moves only 95%).

Oddly, _some_ students are seeing it get stuck. The wheel will stop, but not leave the function. Wiggling the wheel a bit works, so it must have been right on the threshold -- why it only goes 90%, I don't know. Lowering the first threshold to 50% or 75% seems to work, but then again, the 90% threshold worked in extensive tests on my robot. In fact, I had the students send me their code and it worked find on my robot.

Can we make a better MoveFor() routine? Can we get multiple people to test the current one?

**Comment by madhephaestus** [comment]
@gcl8a Ok, pull and test on your end. 

I made 2 changes, first i calculate the position as being withing 2 degres instead of a percent. Then i check for velocities below 10deg/sec where 1 tick in the 50ms loop would produce a 5deg/sec lower bound threshold. 

I also moved the blocking code to its own method, as it may be useful to have just the blocking function.

**Comment by gcl8a** [comment]
Doesn’t Setpoint change during the motion? ie, it moves as a function of time (either liniarly or sinusoid). Wont that make it to the current position reaches the Setpoint before the motor reachies the destination?

**Comment by madhephaestus** [comment]
yeah, thats why the first stage is a while loop checking for the interpolation to be done (you might want to pull again)

**Comment by gcl8a** [comment]
But it breaks out of that while loop almost instantly because Setpoint isn’t the final destination, it’s just a waypoint:

turn
Remaining: 0.82
Remaining: 0.82
Speed: -180.00
Speed: -35.00
Speed: -35.00
Speed: -35.00
Speed: -35.00
Speed: -35.00


Here it breaks out while it’s still along way from the final postion.

**Comment by madhephaestus** [comment]
is this with the current version?

**Comment by gcl8a** [comment]
yes.

I added a counter for the number of times through the first while loop:

0fwd
Remaining: 0.14
Loops: 1
Speed: 30.00
Speed: 30.00
Speed: 30.00
turn
Remaining: 0.57
Loops: 1
Speed: -190.00
Speed: -40.00
Speed: -40.00
Speed: -40.00
Speed: -40.00
Speed: -40.00
turndone


It only does it once (and because tehe next Setpoint hasn’t moved much by interpoation, the error is small).

**Comment by madhephaestus** [comment]
Add all my prints back in:
```
	float distanceToGo;
	// First wait for the interpolation to finish
	while(getInterpolationUnitIncrement()<1){
		delay(10);
		Serial.println(" Interpolation "+String (getInterpolationUnitIncrement()));
	}
	do
	{
		delay(10);
		distanceToGo=fabs((Setpoint*TICKS_TO_DEGREES) - getCurrentDegrees());
		Serial.println("Remaining: "+String(distanceToGo));
	}while (distanceToGo>1.0 );// get within 4 degrees
	// wait for the velocity to be below 10deg/sec
	// 5deg/sec is lower bound of detection
	while (fabs(getDegreesPerSecond()) > 10) {
		Serial.println("Speed: "+String(getDegreesPerSecond()));
		delay(10);
	}
```
the let me know what it does?

**Comment by gcl8a** [comment]
Ah! 

I did not realize that it spent most of the time in the interpolation loop. That makes more sense.

But now it seems like you are doing the same thing twice (once for the interpolation to get to 1 and then again to get to the target). Why not just check for the target? (Which is what I was doing, in a cruder way)

Notes:
I noticed the error in my original code, where I set  int64_t startPos = getCurrentDegrees();   (should be float!!!)
You might want to say while(getInterpolationUnitIncrement()<0.99)  ???

**Comment by madhephaestus** [comment]
Not exactly the same. The first loop blocks until the interpolation is
done. This waits for Setpoint to get to it's final value. Then using
Setpoint, we check that the pid has converged to within 1 degree. Finally
wait for the pid to settle with the velocity check of less than 2 ticks per
50ms sampling of the velocity, aka 10deg/sec.


As for checking for 1 or 0.99, look at the method implementation, it will
explicitly return 1 when the interpolation is done.

So is this resolved?

**Comment by gcl8a** [comment]
Imprecise wording on my part. It seems highly unlikely that the second test could pass without the first. So while they’re not testing the same thing, the end result is the same.

That said, the first test isn’t going to hang the program, so I guess it’s semantics. If it works, it works.

I did some brief testing, but not extensive. Can this be released?

By the way, I don’t see how using int64_t would have caused the previous issue. Losing a fraction of a degree isn’t going to kill the 90% test. I still don’t have an answer for why my code fails on some of the students’ robots. But if this new one can be proved to works, it doesn’t really matter — we have bigger fish to fry..

**Comment by madhephaestus** [comment]
this change is released in 0.2.0
